File: src/main/java/ch/zxseitz/j3de/graphics/scene/components/MeshComponent.java
package ch.zxseitz.j3de.graphics.scene.components;

import ch.zxseitz.j3de.exceptions.J3deException;
import ch.zxseitz.j3de.graphics.mesh.Mesh;
import ch.zxseitz.j3de.graphics.scene.Actor;
import ch.zxseitz.j3de.math.Matrix4;
import ch.zxseitz.j3de.utils.ErrorUtils;

import static org.lwjgl.opengl.GL45.*;

public class MeshComponent implements IComponent {
    private Mesh mesh;

    public MeshComponent(Mesh mesh) {
        this.mesh = mesh;
    }

    @Override
    public void update(double delta, Actor actor, Matrix4 t) throws J3deException {
        if (mesh != null) {
            var program = mesh.getProgram();
            //todo implement scene program to reduce program switch
            program.use();
            //todo generalize
            program.writeUniform("P", actor.getScene().getCamera().getProjection());
            program.writeUniform("T", t);
            var texture = mesh.getTexture();
            if (texture != null) {
                program.writeUniform("tex", texture.id);
            }

            // render mesh
            glBindVertexArray(mesh.getVao());
            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, mesh.getEbo());
            glDrawRangeElementsBaseVertex( mesh.getMode().id, mesh.getStart(), mesh.getEnd(),
                    mesh.count(), GL_UNSIGNED_INT, 0, mesh.getVertexStart());

            var error = glGetError();
            if (error != GL_NO_ERROR) {
                throw new J3deException(ErrorUtils.getErrorInfo(error));
            }
        }
    }
}

File: src/main/java/ch/zxseitz/j3de/graphics/scene/components/MeshComponent.kt
package ch.zxseitz.j3de.graphics.scene.components;

import ch.zxseitz.j3de.J3deException
import ch.zxseitz.j3de.graphics.mesh.Mesh;
import ch.zxseitz.j3de.graphics.scene.Actor;
import ch.zxseitz.j3de.math.Matrix4;
import ch.zxseitz.j3de.utils.ErrorUtils;

import org.lwjgl.opengl.GL45.*;

class MeshComponent(private val mesh: Mesh?): IComponent {
     override fun update(delta: Double, actor: Actor, t: Matrix4) {
        if (mesh != null) {
            val program = mesh.program;
            //todo implement scene program to reduce program switch
            program.use();
            //todo generalize
            program.writeUniform("P", actor.scene.camera.projection);
            program.writeUniform("T", t);
            val texture = mesh.texture;
            if (texture != null) {
                program.writeUniform("tex", texture.id);
            }

            // render mesh
            glBindVertexArray(mesh.vao);
            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, mesh.ebo);
            glDrawRangeElementsBaseVertex( mesh.mode.id, mesh.start, mesh.end,
                    mesh.count(), GL_UNSIGNED_INT, 0, mesh.vertexStart);

            val error = glGetError();
            if (error != GL_NO_ERROR) {
                throw J3deException(ErrorUtils.getErrorInfo(error));
            }
        }
    }
}
